pkg/controller/bypass/bypass_test.go

```diff
@@ -0,0 +1,287 @@
+			beforeFunc: func() {
+				addFuncCallCount = 0
+				delFuncCallCount = 0
+				_, err := client.CoreV1().Pods("default").Create(context.TODO(), podWithLabel, metav1.CreateOptions{})
+				require.NoError(t, err, "failed to create test pod")
+				podWithLabel.DeletionTimestamp = &metav1.Time{Time: time.Now()}
+				_, err = client.CoreV1().Pods("default").Update(context.TODO(), podWithLabel, metav1.UpdateOptions{})
+				require.NoError(t, err, "failed to update test pod with deletion timestamp")
+
+				err = client.CoreV1().Pods("default").Delete(context.TODO(), "test-pod-with-label", metav1.DeleteOptions{})
+				require.NoError(t, err, "failed to delete test pod")
+			},
+			afterFunc: func() {
+				assert.Equal(t, 1, addFuncCallCount, "AddFunc should not been called")
+				assert.Equal(t, 0, delFuncCallCount, "DelFunc should have been called once")
+			},
+			wantErr: false,
+		},
+	}
+
+	go StartByPassController(client)
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tt.beforeFunc()
+			time.Sleep(1 * time.Second)
+			tt.afterFunc()
+		})
+	}
+}
+
+func TestCheckKmeshManaged(t *testing.T) {
+	type args struct {
+		pod *corev1.Pod
+	}
+	tests := []struct {
+		name string
+		args args
+		want bool
+	}{
+		{
+			name: "test1: annotation with kmesh.net/redirection=enabled, should return true",
+			args: args{
+				pod: &corev1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "utPod",
+						Namespace: "utNs",
+						Annotations: map[string]string{
+							"kmesh.net/redirection": "enabled",
+						},
+					},
+				},
+			},
+			want: true,
+		},
+		{
+			name: "test2: annotation without kmesh.net/redirection, should return false",
+			args: args{
+				pod: &corev1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "utPod",
+						Namespace: "utNs",
+						Annotations: map[string]string{
+							"sidecar.istio.io/inject": "true",
+						},
+					},
+				},
+			},
+			want: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := isKmeshManaged(tt.args.pod); got != tt.want {
+				t.Errorf("isKmeshManaged() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestCheckSidecar(t *testing.T) {
+	type args struct {
+		client kubernetes.Interface
+		pod    *corev1.Pod
+	}
+	tests := []struct {
+		name string
+		args args
+		want bool
+	}{
+		{
+			name: "test1: pod withsidecar inject annotation, should return true",
+			args: args{
+				client: fake.NewSimpleClientset(&corev1.Namespace{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "utNs",
+					},
+				}),
+				pod: &corev1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "utPod",
+						Namespace: "utNs",
+						Annotations: map[string]string{
+							"sidecar.istio.io/inject": "true",
+						},
+					},
+				},
+			},
+			want: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got, _ := checkSidecar(tt.args.client, tt.args.pod); got != tt.want {
```

Comment:
what does `client.CoreV1().Namespaces().Get` return, when you did not create a namespace